### ISSUE TITLE
Gless 2 support for Godot 4

### ISSUE THREAD

**Issue by ahmutkadim** [godotengine/godot#77795]
### Godot version

4.0.3

### System information

macOS 13

### Issue description

I am currently using Godot 4 and still Gless 2 renderer is not implemented. I followed through all development builds but none of them included such future to my knowledge. I think this renderer is very important since Apple only supports that in their mobile devices so it is crucial for web games. When will it came?

### Steps to reproduce

Export the project as a web build and try to run it on IOS

### Minimal reproduction project

None

**Comment by mrjustaguy** [comment]
GLES 2 isn't planned in 4.0. see https://godotengine.org/article/about-godot4-vulkan-gles3-and-gles2/

**Comment by AThousandShips** [comment]
> Apple only supports that in their mobile devices

Can you provide a source for this?

**Comment by clayjohn** [comment]
iOS supports OpenGL ES 3.0 https://developer.apple.com/documentation/opengles

What iOS doesn't yet support is WebGL2 which is needed to run Godot 4 on the web. iOS only supports WebGL which corresponds to OpenGL ES 2.0.

**Comment by ahmutkadim** [comment]
I may have phrased my problem wrong. Yes OpenGL might be supported in new versions in IOS but it still does not help. Apple might support different rendering techniques as well but still my testings indicate that my Godot games made in Godot 4 cannot be opened in any IOS device. I have tried various IOS devices but it does not work. However in mobile Godot 3 worked since I was using Gless 2. What may I do now ?

**Comment by AThousandShips** [comment]
As said above the issue is WebGL2, are you unable to export for iOS instead of for web?

**Comment by AThousandShips** [comment]
Any reason you have to export for web and not directly for iOS?

**Comment by ahmutkadim** [comment]
It is costly and hard to publish an app on App Store and I prefer publishing my games on web since it is more accessible for all users. But this rendering problem hinders such accessibility.

**Comment by AThousandShips** [comment]
That's understandable, but unfortunately this is a limitation on the side of Apple, and as the article above on the reasons for this says there's a lot of reasons for this decision

**Comment by ahmutkadim** [comment]
Okay I understand.

**Comment by AlexBowring** [comment]
The response to this issue speaks volumes. Obviously people are going to be using the Godot engine to make games that they want to be accessible to everyone. If, because of the renderer Godot is using, those games can't be played on apple devices (the most popular tech company _in the world_), then surely its in Godot's best interest to themselves as well as their users to provide a solution to ensure the games are accessible on apple devices, rather then say its a problem of Apples. This sort of thing is a real put off, and its even more frustrating that this problem didn't exist in a previous Godot version but does exist in a newer version.

**Comment by lochana-dineth** [comment]
Being Pragmatic is the best here. I myself can't run Godot 4 due to a weird quirk of the Intel HD 3000 on WIndows. So, I am using Godot 3 for now. Is it good, no.
Then again if there is no support for it in Godot4, no one has interest in implementing it. It is what it is.

**Comment by causal-guide** [comment]
we need it

**Comment by Naming-things-is-hard-btw** [comment]
well how hard can this be to implement? i mean.. i made some [terrible] 3d renderers in c++ in the past and im willing to put my time into this since most android devices support gles3 but very poorly so on a phone gles2 is a must

ill try to help with this but i cant promise much...

**Comment by AThousandShips** [comment]
It's not just a question of how difficult it would be to implement, but how difficult it would be to support a *fourth* rendering mode and set of shaders and workflows, having two backends for compatibility would add a lot more maintenance work

The compatibility backend with the features of gles3 is already behind and cut down, going even lighter and more simplified and less powerful with gles2 might be too far to justify the work, would it not be better to use Godot 3.x if you need to target such devices? Could they even run the rest of the Godot engine? CPU power, ram, storage, etc.?

**Comment by Naming-things-is-hard-btw** [comment]
> It's not just a question of how difficult it would be to implement, but how difficult it would be to support a _fourth_ rendering mode and set of shaders and workflows, having two backends for compatibility would add a lot more maintenance work
> 
> The compatibility backend with the features of gles3 is already behind and cut down, going even lighter and more simplified and less powerful with gles2 might be too far to justify the work, would it not be better to use Godot 3.x if you need to target such devices? Could they even run the rest of the Godot engine? CPU power, ram, storage, etc.?
> 
> Like, realistically, what is the market for:
> 
>     * Devices that doesn't support gles3
> 
>     * Has enough CPU and memory to run Godot 4.x well enough to justify using Godot 4.x
> 
> 
> The main reason I'd say to run Godot 4 on a device like this is to be able to run more advanced stuff on other devices, but if you just want the power and features that can run on such a device I'd say Godot 3.x is a far better choice



but if a devoloper wanted to make a simple game (2d game for example) they would want it to run on all devices that can render 2d right?

maybe a better solution would be is to make the rendering part in godot extendable via plugins? so if someone wants gles2 they just make a plugin for that?

**Comment by AThousandShips** [comment]
> but if a devoloper wanted to make a simple game (2d game for example) they would want it to run on all devices that can render 2d right?

Then Godot 3.x would be better suited I'd say, if you don't want all the advanced features in 4.x, 4.x takes more performance even when not using all the extra stuff because of basic limitations 

4.x has higher minimum requirements to allow for higher end features and performance, it's hard to have very low and very high levels in one engine

**Comment by Naming-things-is-hard-btw** [comment]
> > but if a devoloper wanted to make a simple game (2d game for example) they would want it to run on all devices that can render 2d right?
> 
> Then Godot 3.x would be better suited I'd say, if you don't want all the advanced features in 4.x, 4.x takes more performance even when not using all the extra stuff because of basic limitations
> 
> 4.x has higher minimum requirements to allow for higher end features and performance, it's hard to have very low and very high levels in one engine

well yea i kinda agree with this but isnt backporting featues (especially gdscript features) harder than having a third renderer?
there are alot of bugs fixed in godot 4 but are still there in the newest godot 3.6 releases

**Comment by Naming-things-is-hard-btw** [comment]
> > but if a devoloper wanted to make a simple game (2d game for example) they would want it to run on all devices that can render 2d right?
> 
> Then Godot 3.x would be better suited I'd say, if you don't want all the advanced features in 4.x, 4.x takes more performance even when not using all the extra stuff because of basic limitations
> 
> 4.x has higher minimum requirements to allow for higher end features and performance, it's hard to have very low and very high levels in one engine

with having a gles 2 in godot 4 there will be no need to mantain godot 3 at all

**Comment by AThousandShips** [comment]
So then my question remains:
How large is this potential market for devices that don't support gles3 but can run Godot 4?

> with having a gles 2 in godot 4 there will be no need to mantain godot 3 at all

Not that simple, as I said, Godot 3 has very different hardware and performance requirements

**Comment by Naming-things-is-hard-btw** [comment]
> So then my question remains: How large is this potential market for devices that don't support gles3 but can run Godot 4?
> 
> > with having a gles 2 in godot 4 there will be no need to mantain godot 3 at all
> 
> Not that simple, as I said, Godot 3 has very different hardware and performance requirements

i might have worded it poorly [sorry]
im talking about phones not desktop, we cant ignore the fact that phones are alot slower in rendering power than cpu power they can handle cpu heavy tasks but even a simple shader completely kills the device

i did some tests and on my phone godot 4 runs flawlessly when we talk about cpu usage but it struggles to render things

**Comment by Naming-things-is-hard-btw** [comment]
> So then my question remains: How large is this potential market for devices that don't support gles3 but can run Godot 4? If there is that's a valid argument that this is something to consider, otherwise there's no point at all to support gles2
> 
> > with having a gles 2 in godot 4 there will be no need to mantain godot 3 at all
> 
> Not that simple, as I said, Godot 3 has very different hardware and performance requirements

since you say that gles2 isnt important then why all unity games support it

**Comment by AThousandShips** [comment]
> since you say that gles2 isnt important

I didn't say anything like that, take a moment to read what I'm saying again

**Comment by Naming-things-is-hard-btw** [comment]
> So then my question remains: How large is this potential market for devices that don't support gles3 but can run Godot 4? If there is that's a valid argument that this is something to consider, otherwise there's no point at all to support gles2
> 
> > with having a gles 2 in godot 4 there will be no need to mantain godot 3 at all
> 
> Not that simple, as I said, Godot 3 has very different hardware and performance requirements

heres the point u said "How large is this potential market for devices that don't support gles3" but im not talking about devices that doesnt support gles3 im talking about the huge list of devices that (fake) supports it that its filled with bugs and and performance issues that its not usable

**Comment by AThousandShips** [comment]
Yes, that's what "doesn't support gles3" means, I'm sorry I thought it was obvious, that isn't supporting gles3, it's *claiming* to support it

**Comment by Zireael07** [comment]
@AThousandShips In that case the list is pretty big - most Mali and Adreno mobile GPUs *claim* to support but do not really resulting in endless bugs and/or poor performance

**Comment by AThousandShips** [comment]
A lot of those are being worked on, but that was the primary question, but the central point remain:
Adding gles2 would involve a lot of extra work and extra maintenance for a feature targeting a relatively small niche of targets probably better served by 3.x

It's not simply a matter of adding some code and being done, the gles3 renderer takes a lot of maintenance already, especially with driver capacity etc., including handling the weird quirks of some drivers

**Comment by Naming-things-is-hard-btw** [comment]
welp... i guess were back to unity :/
